libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TemplatesRegisterExtensions.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using AdaptiveExpressions;
+using AdaptiveExpressions.Memory;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Generators;
+using Microsoft.Bot.Builder.LanguageGeneration;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// Register lg templates into Expression.
+    /// Using prefix lg. to avoid the conflict with pre-built functions.
+    /// </summary>
+    public static class TemplatesRegisterExtensions
+    {
+        /// <summary>
+        /// Register Template functions.
+        /// </summary>
+        /// <param name="dialogContext">Dialog context.</param>
+        /// <param name="generator">Language generator.</param>
+        public static void RegisterTemplateFunctions(this DialogContext dialogContext, LanguageGenerator generator)
+        {
+            var lgm = dialogContext.Services.Get<LanguageGeneratorManager>();
+            var templateNames = GetTemplates(generator, dialogContext).Distinct();
+
+            var prebuildFunctionNames = Expression.Functions.Values.Select(u => u.Type);
+            foreach (var templateName in templateNames)
+            {
+                var functionType = prebuildFunctionNames.Contains(templateName) ? "lg." + templateName : templateName;
```

Comment:
>"lg." [](http://example.com/codeflow?start=82&length=5)

why wouldn't you just use the resource.id as the namespace?
"foo.lg.TemplateX()" => would bind to foo-{lang}.lg via fallbacks, and you know the resource.id (so no need save off the resource Id or anything.
This is what I thought your PR was going to do.

Reply:
I didn't register it with resource name prefix, because i want all actions have the same access to LG templates. So if in SendActivity i don't have to call my templates with foo.lg.T1, then in all other actions, user should not have to do that either. Does that make sense to you?